### PR TITLE
SI-5639 Fix spurious discarding of implicit import

### PR DESCRIPTION
**Pull request by retronym**
In Scala fa0cdc7b (just before 2.9.0), a regression in implicit search,
SI-2866, was discovered by Lift and fixed. The nature of the regression was
that an in-scope, non-implicit symbol failed to shadow an eponymous
implicit import.

The fix for this introduced `isQualifyingImplicit` which discards in-scope
implicits when the current `Context`'s scope contains a name-clashing
entry.

Incidentally, this proved to be a shallow solution, and we later improved
shadowing detection in SI-4270 / 9129cfe9. That picked up cases where a
locally defined symbol in an intervening scope shadowed an implicit.

This commit includes the test case from the comments of SI-2866. Part of it
is tested as a neg test (to test reporting of ambiguities), and the rest is
tested in a run test (to test which implicits are picked.)

However, in the test case of SI-5639, we see that the scope lookup
performend by `isQualifyingImplicit` is fooled by a "ghost" module symbol.
The apparition I'm referring to is entered when
`initializeFromClassPath` / `enterClassAndModule` encounters a class file
named 'Baz.class', and speculatively enters _both_ a class and module
symbol. AFAIK, this is done to defer parsing the class file to determine
what inside. If it happens to be a Java compiled class, the module symbol
is needed to house the static members.

This commit adds a condition that `Symbol#exists` which shines a torch
(forces the info) in the direction of the ghost module symbol. In our test,
this causes it to vanish, as we only need a class symbol for the Scala
defined `class Baz`.

The existing `pos` test for this actually did not exercise the bug,
separate compilation is required. It was originally checked in to
`pending` with this error, and then later moved to `pos` when someone
noticed it was not failing.

To be conservative, I've predicated this fix in `-Xsource:2.12`. This is
done in separate commit to show that the previous fix passes the test
suite, rather than just tests with `-Xsource:2.12`.

Review by @adriaanm

### CODE REVIEW

File: src/compiler/scala/tools/nsc/typechecker/Contexts.scala [hunk 1]
```diff
@@ -798,7 +798,7 @@ trait Contexts { self: Analyzer =>
       isAccessible(sym, pre) &&
       !(imported && {
         val e = scope.lookupEntry(name)
-        (e ne null) && (e.owner == scope)
+        (e ne null) && (e.owner == scope) && (e.sym.exists || !settings.isScala212)
```
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/typechecker/Contexts.scala]: should we swap the order of the operands to `||` to avoid shining our torch too eagerly?

**Comment by adriaanm** [comment]
I really like that you included a run test! Very thorough testing and great analysis.

File: src/compiler/scala/tools/nsc/typechecker/Contexts.scala (hunk 1 above)
Comment (retronym) [inline src/compiler/scala/tools/nsc/typechecker/Contexts.scala]: done.

**Comment by retronym** [comment]
@adriaanm I switched the order as suggested, r-r-r-r-ready for re-review.

**Comment by adriaanm** [comment]
LGTM!

**Pull request merged**